Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Hole_filling/Triangulate_hole_polyline.h

```diff
@@ -1210,16 +1210,87 @@ class Triangulate_hole_polyline {
  ************************************************************************/
 // /!\ points.first == points.last
 
+template <typename Traits>
+bool is_planar_2(
+  const std::vector<typename Traits::Point_3>& points,
+  const typename Traits::Vector_3& avg_normal,
+  const typename Traits::FT max_distance,
+  const Traits& traits) {
+
+  typedef typename Traits::FT FT;
+  typedef typename Traits::Point_3 Point_3;
+  typedef typename Traits::Plane_3 Plane_3;
+  typedef typename Traits::Compute_squared_distance_3 Squared_distance_3;
+
+  const Squared_distance_3 squared_distance_3 =
+    traits.compute_squared_distance_3_object();
+
+  const std::size_t n = points.size() - 1; // the first equals to the last
+  if (n < 3)
+    return false; // cant be a plane!
+
+  // Compute centroid.
+  FT cx = FT(0), cy = FT(0), cz = FT(0);
+  for (std::size_t i = 0; i < n; ++i) {
+    const Point_3& p = points[i];
+    cx += p.x();
+    cy += p.y();
+    cz += p.z();
+  }
+  cx /= static_cast<FT>(n);
+  cy /= static_cast<FT>(n);
+  cz /= static_cast<FT>(n);
+  const Point_3 centroid = Point_3(cx, cy, cz);
+
+  // Compute covariance matrix.
+  FT xx = FT(0), yy = FT(0), zz = FT(0);
+  FT xy = FT(0), xz = FT(0), yz = FT(0);
+  for (std::size_t i = 0; i < n; ++i) {
+    const Point_3& p = points[i];
+    const FT dx = p.x() - cx;
+    const FT dy = p.y() - cy;
+    const FT dz = p.z() - cz;
+    xx += dx * dx; yy += dy * dy; zz += dz * dz;
+    xy += dx * dy; xz += dx * dz; yz += dy * dz;
+  }
+
+  // Check the planarity.
+  const FT x = yy * zz - yz * yz;
+  const FT y = xx * zz - xz * xz;
+  const FT z = xx * yy - xy * xy;
+  FT maxv = -FT(1);
+  maxv = (CGAL::max)(maxv, x);
+  maxv = (CGAL::max)(maxv, y);
+  maxv = (CGAL::max)(maxv, z);
+  if (maxv <= FT(0))
+    return false; // a good plane does not exist for sure!
+
+  const Plane_3 plane = Plane_3(centroid, avg_normal);
+  FT avg_sq_distance = FT(0);
+  for (std::size_t i = 0; i < n; ++i) {
+    const Point_3& p = points[i];
+    const Point_3  q = plane.projection(p);
```

Comment:
@sloriot @janetournois Is it ok that constructors and member functions are used instead of kernel functors?

Reply:
since the traits is a parameter of the function, we can use the traits to get the functors

Reply:
I will fix that.